cylc/uiserver/app.py

```diff
@@ -0,0 +1,369 @@
+            self.workflows_mgr.update
+        )
+        ioloop.PeriodicCallback(
+            self.workflows_mgr.update,
+            self.scan_interval * 1000
+        ).start()
+
+    def initialize_settings(self):
+        """Update extension settings.
+
+        Update the self.settings trait to pass extra settings to the underlying
+        Tornado Web Application.
+
+        self.settings.update({'<trait>':...})
+        """
+        super().initialize_settings()
+        self.log.info("Starting Cylc UI Server")
+        self.log.info(f'Serving UI from: {self.ui_path}')
+
+    def initialize_handlers(self):
+        self.handlers.extend([
+            ('cylc/hello', MyExtensionHandler),
+            (
+                'cylc/graphql',
+                UIServerGraphQLHandler,
+                {
+                    'schema': schema,
+                    'resolvers': self.resolvers,
+                    'backend': CylcGraphQLBackend(),
+                    'middleware': [IgnoreFieldMiddleware],
+                }
+            ),
+            (
+                'cylc/graphql/batch',
+                UIServerGraphQLHandler,
+                {
+                    'schema': schema,
+                    'resolvers': self.resolvers,
+                    'backend': CylcGraphQLBackend(),
+                    'middleware': [IgnoreFieldMiddleware],
+                    'batch': True,
+                }
+            ),
+            (
+                'cylc/subscriptions',
+                SubscriptionHandler,
+                {
+                    'sub_server': self.subscription_server,
+                    'resolvers': self.resolvers,
+                }
+            ),
+            (
+                'cylc/userprofile',
+                UserProfileHandler,
+            ),
+            (
+                'cylc/(.*)?',
+                # web.StaticFileHandler,
+                CylcStaticHandler,
+                {
+                    'path': str(self.ui_path),
+                    'default_filename': 'index.html'
+                    # 'url': url_path_join(
+                    #     self.static_url_prefix, 'index.html'
+                    # )
+                }
+            ),
+            (
+                # redirect '/cylc' to '/cylc/'
+                'cylc',
+                RedirectHandler,
+                {
+                    'url': 'cylc/'
+                }
+            )
+        ])
+
+    def initialize_templates(self):
+        """Change the jinja templating environment."""
+
+    @classmethod
+    def launch_instance(cls, argv=None, **kwargs):
+        if argv is None:
+            # jupyter server isn't expecting to be launched by a Cylc command
+            # this patches some internal logic
+            argv = sys.argv[2:]
+        super().launch_instance(argv=argv, **kwargs)
+
+    async def stop_extension(self):
+        for sub in self.data_store_mgr.w_subs.values():
+            sub.stop()
```

Comment:
Is `stop_extension` stopping the Python executable? I think it is, but asking just in case it's not. If not, I think we need to delete from the dictionary the stopped subs.

Also, should we wrap in a try/catch these calls if terminating program? To make sure we try to stop, shutdown, and destroy?

Reply:
The `stop_extension` hook allows extensions to clean up any resources (e.g. subs) they have created before the server shuts down and the Python process exits.

We need to stop the subs as otherwise the server will hang rather than shutting down.